### ISSUE TITLE
Don't mutate existing validations and validation plans - create new ones!

### ISSUE THREAD

**Issue by karenetheridge** [joyent/conch#546]
Existing `validation_plan` and `validation` rows in the database should never be mutated -- or historical data (`validation_state`, linked to `device_report`) cannot be used to give an accurate picture of what checks were done on a device.

If a validation db row is found to not match a validation module on disk, deactivate it, remove it from all validation_plans, create a new validation row, deactivate all validation_plans containing that old validation, and create an equivalent one with the new validation as a member.

(Deactivating validation plans and creating new ones should be done very carefully, only with the explicit consent of the operator. Therefore only do this by execution of a command line utility, not as some side effect in the application itself.)

Builds on work done for #545.

**Comment by karenetheridge** [comment]
This will also include removing the endpoints:

    POST /validation_plan/:validation_plan_id/validation
    DELETE /validation_plan/:validation_plan_id/validation/:validation_id

and adding:

    DELETE /validation_plan/:validation_plan_id

..which requires adjustments in the shell.

**Comment by sungo** [comment]
Confused. Are you proposing that we remove the ability for an operator, at runtime, to modify and delete validation plans?

**Comment by karenetheridge** [comment]
Yes I am -- see the first sentence of this issue. IMO validations and plans should be immutable once created, or we can't track history accurately.

**Comment by sungo** [comment]
After talking it through in chat, here's where I'm at currently:

The _right_ feature-set is for an operator to be able to create, remove, modify validation plans at whim

However:
* The current code base uses static dispatch and only allows two validation plans to be used, changing between them based on a field in the device report. This static dispatch requires that two plans exist with predetermined names.
* Validation code is stored on disk using a string name and numeric version and are loaded into the database at run time and no guaranteed UUIDs. 

So we have a validation system that depends on operators creating a plan with a known string name that contains validation UUIDs that happen to correspond to the ones that an API node created when it started up.

The system we have is ... not what we need. 

I have come around to the notion that we should ponder removing the API endpoints for allowing the CRUD of the validation system. But *only* in the context of fixing the system and re-introducing the feature set later. 

However, we have a chicken/egg problem. In the current codebase, we need those plans with those specific names in order to run validations at all. #545 proposes that we stop inserting validations at runtime. We removed the code that built validation plans at run time months ago. Thus, when one installs Conch, the system can do nothing at all with device reports. The system does not automatically create anything and we'll have removed the API endpoints for an operator to fix that on their own. 

What are your thoughts on tackling that issue?

**Comment by karenetheridge** [comment]
Right now (in today's master), if an operator spun up a brand new instance of conch, they would have no validation plans in the database, but they would get (via `Conch::ValidationSystem->load_validations()`, run at application startup) all the validation records created from Conch::Validation::* modules -- all they need to do is create the two hardcoded plans that Conch::Controller::DeviceReport expects.

This is not a great situation, but it's also not horrible either.  I can fairly easily write a command-line script `bin/conch whatever` that adds the expected validation_plans, that would get the new database into alignment with what we are currently running in production.  The operator can also make use of the existing `POST /validation_plan/...` endpoints to set things up as he wishes.

My original thoughts in this issue were to make validation_plans and validations less mutable, choosing to deactivate/clone/create rather than edit in place.  The piece that is missing is the hardcoded device type -> validation plan name dispatch. We can flesh that out now with.a query param/Content-Type header dispatching system, or let it sit as is until the device lifecycle re locker is fleshed out a little more.

We also discussed the possibility of ripping out all the validation_plan routes in Conch::Route::Validation. That's an easier option for me, as deleting code is easy.

Another option is doing nothing, i.e. acting on nothing described in this ticket until a little later down the road.  That's also easy.

**Comment by sungo** [comment]
I think that the unfortunate but correct path right now is:
* As you say, a mojo command `carton exec bin/conch wtf` for creating the named validation plans that Joyent expects, how Joyent expects them, and document it in the README. Particularly since failure to set all that up makes a default instance pretty useless.
* Remove the API endpoints for CRUDing validation structures, for now

We can look to content type dispatch (as in the `sungo/validations-by-content-type` branch) and locker stuff elsewhere. 

Basically, I think we should lock down the current validation system until we can properly address the concerns we have about the engine itself and the future of the subsystem re locker

**Comment by karenetheridge** [comment]
#545 is merged.  Shall I disable the POST and DELETE endpoints for `/validation/*` and `/validation_plan/*` for now?

**Comment by sungo** [comment]
I vote yes.

**Comment by karenetheridge** [comment]
Validations are no longer mutable in api endpoints. Putting this on hold until we figure out what we want to do next.

**Comment by karenetheridge** [comment]
We'll need this in v2 if someone ever needs to alter an existing validation or add a new one, since there are no longer endpoints for operating on validation plans. I'll write a command line utility that upgrades validation plans when their validation members change (the idea being that this utility is run immediately after deploying a code update that contains alterations to validations).
